### PR TITLE
Fixes for grayscale conversion

### PR DESCRIPTION
**Pull request by peterwilliams97**
Preserve the SMask in an image XObject when converting the image XObject to grayscale.

This requires converting that Matte in the masking image to grayscale if it has a Matte.

[smask.pdf](https://github.com/unidoc/unidoc/files/1969783/smask.pdf)
 will convert correctly with this fix. 

You will need the corresponding change to unidoc-examples that I making a pull request for now.

### CODE REVIEW

**Comment by gunnsth** [comment]
Nice work Peter.  The only thing that needs to be fixed is:

We are following a semantic versioning scheme, so
-func (this *PdfTilingPattern) GetContentStream() ([]byte, error) {
+func (this *PdfTilingPattern) GetContentStream() ([]byte, StreamEncoder, error) {

is a breaking change.  Is it necessary to return the encoder in this case?

Otherwise, I would suggest adding a new function called
+func (this *PdfTilingPattern) GetContentStreamWithEncoder() ([]byte, StreamEncoder, error) {
...
and making the GetContentStream very basic,
just stream, _, err := GetContentStreamWithEncoder(...)
and return stream, err...
and a comment above: // TODO (v3): Change GetContentStreamWithEncoder to GetContentStream

**Comment by peterwilliams97** [comment]
Done

File: pdf/model/pattern.go [hunk 1]
```diff
@@ -81,21 +81,34 @@ func (this *PdfTilingPattern) IsColored() bool {
 	}
 }
 
-// Get the pattern cell's content stream.
+// GetContentStream returns the pattern cell's content stream
 func (this *PdfTilingPattern) GetContentStream() ([]byte, error) {
+	decoded, _, nil := this.GetContentStreamWithEncoder()
+	return decoded, nil
```
Comment (gunnsth) [inline pdf/model/pattern.go]: Should probably return the error if returned by GetContentStreamWithEncoder ?

File: pdf/model/xobject.go [hunk 2]
```diff
@@ -281,11 +292,74 @@ func NewXObjectImageFromImage(img *Image, cs PdfColorspace, encoder StreamEncode
...
 	return xobj, nil
 }
 
+// smaskMatteToGray converts to gray the Matte value in the SMask image referenced by `xobj` (if
+// there is one)
+func smaskMatteToGray(xobj *XObjectImage) error {
+	if xobj.SMask == nil {
+		return nil
+	}
+	stream := xobj.SMask.(*PdfObjectStream)
```
Comment (gunnsth) [inline pdf/model/xobject.go]: Might be worth type checking, e.g.
stream, ok := xobj.SMask.(*PdfObjectStream)
if !ok {
  return ErrTypeCheck
}

**Pull request merged**